runtime/scripts/math.js

```diff
@@ -1972,6 +1972,28 @@ var math = Numbas.math = /** @lends Numbas.math */ {
             product = math.mul(product, list[i]);
         }
         return product;
+    },
+    /** Weighted random given a list of [item, probability] 
+     * 
+     * @param {Array} list
+     * @returns {number}
+    */
+     weighted_random: function(list) {
```

Comment:
You've used a few ES6 features -  `Map`, list destructuring and `for .. of ..` - which we can't use yet because we're still supporting old browsers - see #527.

There's no need for a Map here anyway: you iterate over the partial sums in the same order they appear in the list, so you might as well just iterate over the input list twice.